### PR TITLE
feat(dotfiles): reconcile profile source switches

### PR DESCRIPTION
**Pull request by jdx**
## Summary
- reconcile stale `symlink-each` state when an environment/profile changes the source for a target
- remove only links still pointing at the previously recorded source
- preserve unmanaged paths and paths desired by another active source
- preview stale-link removals in dry-run mode

## Testing
- `cargo check --locked`
- `mise run test:e2e e2e/cli/test_dotfiles_files`
- `mise run lint-fix`

Stacked on #12523.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

---

> [!NOTE]
> **Medium Risk**
> Changes dotfiles apply/unlink behavior on shared targets and mutates filesystem state from persisted history; logic is conservative (only removes links still pointing at recorded sources) but mistakes could delete wrong symlinks.
> 
> **Overview**
> **`symlink-each` apply now reconciles prior profile state** when the same target is managed by a different source (e.g. switching `MISE_ENV` between home and work). Apply scans stored ownership under `$MISE_STATE_DIR/dotfiles`, removes links that still point at a **superseded** source but are no longer desired, repoints shared paths, and leaves unmanaged files alone. Dry-run prints those `rm` steps; empty dirs from removed links are cleaned up afterward.
> 
> Path handling is tightened: **lexical normalization** is used for state files, recorded links, and `link_points_to` so equivalent spellings (`./profile-home` vs `profile-home`) match. **`plan_apply_with_active`** lets apply plan against the config that will be active after a transaction—`dotfiles add` uses this so re-adding an already-applied `symlink-each` entry plans from the configured source, not the live target used for capture validation.
> 
> Docs describe profile switching; e2e covers home/work toggling and add refresh; unit tests cover reconciliation edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6f10eb505d61196237ea94f161a3dead2db2c99d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/system/files.rs [hunk 1]
```diff
@@ -1077,6 +1082,79 @@ fn load_symlink_each_state(req: &FileRequest) -> LoadedSymlinkEachState {
...
+    let mut targets = indexmap::IndexSet::new();
+    for entry in state_dir.read_dir()? {
+        let path = entry?.path();
+        let Ok(state) = read_symlink_each_state(&path) else {
+            continue;
+        };
+        if !valid_symlink_each_state(&state)
+            || !active_targets.contains(&state.target)
+            || active_keys.contains(&(&state.source, &state.target))
+        {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs]: Fixed in 797e9fb90. Reconciliation now derives the desired union from every active configured contributor for each selected target, while the apply work itself remains filtered. A focused unit test covers preserving an omitted active contributor.\n\n*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

File: src/system/files.rs [hunk 2]
```diff
@@ -1077,6 +1082,79 @@ fn load_symlink_each_state(req: &FileRequest) -> LoadedSymlinkEachState {
...
+            continue;
+        }
+        for link in &state.links {
+            if desired_links
+                .get(&link.target)
+                .is_none_or(|source| **source != link.source)
+                && link_points_to(&link.source, &link.target)
+            {
+                stale_links.insert(link.target.clone(), link.clone());
+            }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs:1213]: Fixed in 797e9fb90. Symlink-each state identities, stored source/target paths, and managed link paths are lexically normalized. The profile e2e switches back using ./profile-home and verifies the links converge correctly.\n\n*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

File: src/system/files.rs [hunk 3]
```diff
@@ -1376,14 +1472,25 @@ pub(crate) fn execute_apply(plan: ApplyPlan<'_>, opts: &ApplyOpts) -> Result<boo
...
-            .iter()
-            .map(|(r, _)| r.target_raw.clone())
-            .collect::<Vec<_>>()
-            .join(", ")
-    );
+    cleanup_reconciled_directories(&plan.reconciliation)?;
+    for path in &plan.reconciliation.obsolete_state_paths {
+        if path.exists() {
+            file::remove_file(path)?;
+        }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs]: Fixed in 797e9fb90. Reconciled profile state is retained as historical ownership data instead of deleted, so switching back uses the recorded footprint and does not fall back to recursively walking the target. The e2e now exercises home to work to home while preserving unmanaged content.\n\n*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

File: src/system/files.rs [hunk 4]
```diff
@@ -1091,6 +1100,102 @@ fn load_symlink_each_state(req: &FileRequest) -> LoadedSymlinkEachState {
...
+            targets: vec![],
+        };
+    }
+    let mut stale_links = IndexMap::<PathBuf, ManagedLink>::new();
+    let mut targets = indexmap::IndexSet::new();
+    for state in stored {
+        if !valid_symlink_each_state(state)
+            || !selected_targets.contains(&state.target)
+            || active_keys.contains(&(&state.source, &state.target))
+        {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs:1202]: Fixed in 418fba96c. Persisted symlink-each state is lexically normalized immediately after deserialization, including source, target, and every managed link, before validation or reconciliation. Added a regression unit test for legacy non-normalized state.\n\n*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

File: src/system/files.rs [hunk 5]
```diff
@@ -1091,6 +1100,113 @@ fn load_symlink_each_state(req: &FileRequest) -> LoadedSymlinkEachState {
...
+    let selected_targets = selected_requests
+        .iter()
+        .map(|req| lexical_normalize(&req.target))
+        .collect::<std::collections::HashSet<_>>();
+    let active = active_requests
+        .iter()
+        .filter(|req| {
+            req.mode == FileMode::SymlinkEach
+                && req.source.is_dir()
+                && selected_targets.contains(&lexical_normalize(&req.target))
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs:1155]: Fixed in a2b75d7cc. Active symlink-each declarations now retain a normalized ownership identity even when their source directory is temporarily missing, so reconciliation preserves their stored links while another contributor is selected. Added a regression unit test for this case.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

File: src/system/files.rs [hunk 6]
```diff
@@ -1091,6 +1100,113 @@ fn load_symlink_each_state(req: &FileRequest) -> LoadedSymlinkEachState {
...
+
+fn normalize_symlink_each_state(mut state: SymlinkEachState) -> SymlinkEachState {
+    state.source = lexical_normalize(&state.source);
+    state.target = lexical_normalize(&state.target);
+    for link in &mut state.links {
+        link.source = lexical_normalize(&link.source);
+        link.target = lexical_normalize(&link.target);
+    }
+    state
+}
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs:1133]: Fixed in ce8e738c7. Managed-link verification now lexically resolves the stored symlink destination, including relative destinations, before comparing it with the normalized source. This still recognizes ownership after the source disappears. Added a Unix regression test using a dangling symlink with a non-normalized source path.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

File: e2e/cli/test_dotfiles_files [hunk 7]
```diff
@@ -210,6 +210,39 @@ assert_contains "mise dotfiles status ~/.local/tracked-home" "stale link"
...
+assert "readlink ~/.local/profile-home/shared" "$PWD/profile-work/shared"
+assert "readlink ~/.local/profile-home/work-only" "$PWD/profile-work/work-only"
+assert_fail "test -e ~/.local/profile-home/home-only"
+assert "cat ~/.local/profile-home/unmanaged" "unmanaged"
+# Keep prior profile state so switching back uses the recorded footprint, and
+# normalize equivalent source spellings when identifying that state.
+sed -i.bak 's/source = "profile-home"/source = ".\/profile-home"/' mise.home.toml
+assert_succeed "MISE_ENV=home mise dotfiles apply ~/.local/profile-home --yes"
+assert "readlink -f ~/.local/profile-home/shared" "$PWD/profile-home/shared"
+assert "readlink -f ~/.local/profile-home/home-only" "$PWD/profile-home/home-only"
```
[the remark this answers is not in the document]
Comment (jdx) [reply e2e/cli/test_dotfiles_files]: Fixed in ef32eaaa3. The switch-back assertions now use portable plain readlink and compare the raw target, including the intentionally preserved ./ source spelling. The focused dotfiles E2E passes with that behavior.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

File: src/system/files.rs [hunk 8]
```diff
@@ -1579,12 +1741,40 @@ pub(crate) fn plan_apply<'a>(
     if !problems.is_empty() {
         bail!("files: {}", problems.join("\nfiles: "));
     }
+    let active_requests = files_from_config(config)?;
     Ok(ApplyPlan {
         todo,
         record_symlink_each,
+        reconciliation: plan_symlink_each_reconciliation(&active_requests, requests)?,
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs]: Fixed in 582eed5b1. DotfilesAdd now plans reconciliation against the prospective post-write request set, so a selected re-added symlink-each declaration remains active and retained links are not classified as obsolete. The focused reconciliation tests and dotfiles e2e suite pass.\n\n*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

File: src/cli/dotfiles/add.rs [hunk 9]
```diff
@@ -343,9 +343,10 @@ impl DotfilesAdd {
                 yes: true,
             };
             let apply_plan = if !self.no_apply && !apply_requests.is_empty() {
-                Some(system::files::plan_apply(
+                Some(system::files::plan_apply_with_active(
                     &config,
                     &apply_requests,
+                    &prospective,
                     &apply_opts,
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/cli/dotfiles/add.rs]: Fixed in 6f10eb505. Validation still uses the live target to model the captured filesystem shape, but reconciliation now builds a separate post-write active request set from each accepted item’s configured managed request. Added an e2e regression that refreshes an applied symlink-each entry and verifies its retained link survives.\n\n*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
